### PR TITLE
Patch `ORTTrainer` inference with ONNX Runtime backend

### PR DESCRIPTION
**Pull request by JingyaHuang**
# What does this PR do?

- [x] Enable loss as output in `ORTModelForCausalLM`
- [x] Enable using cache for causal lm task when doing inference in `ORTTrainer`
- [x] Fix the ORT inference of `ORTSeq2SeqTrainer` when `predict_with_generate=True` 

The PR is tested locally, and a PR will be opened for improving the current nightly test.

The caveat #719 always exists, but I won't deal with it because of a lack of bandwidth and need from the community.

### CODE REVIEW

File: optimum/onnxruntime/base.py [hunk 1]
```diff
@@ -195,19 +195,19 @@ def forward(
...
         past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None,
+        labels: Optional[torch.LongTensor] = None,
     ) -> CausalLMOutputWithCrossAttentions:
+        known_output_shapes = {}
         # Flatten the past_key_values
         if past_key_values is not None:
             past_key_values = [past_key_value for pkv_per_layer in past_key_values for past_key_value in pkv_per_layer]
 
         if self.device.type == "cuda" and self.parent_model.use_io_binding:
-            past_key_values_shapes = self.compute_past_key_values_output_shapes(
```
Comment (fxmarty) [inline optimum/onnxruntime/base.py:204]: why this variable renaming?

Comment (JingyaHuang) [reply optimum/onnxruntime/base.py:204]: Because I then added the shape information of loss inside.

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 2]
```diff
@@ -774,6 +774,9 @@ def forward(
             encoder_outputs = self.encoder(input_ids=input_ids, attention_mask=attention_mask)
 
         # Decode
+        if decoder_input_ids is None:
+            decoder_input_ids = input_ids
+
```
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_seq2seq.py]: Replaced by raising an error as transformers does.

**Review by fxmarty — APPROVED** [review]
LGTM, thank you!
